### PR TITLE
Track deletion is handled by a service

### PR DESCRIPTION
**Pull request by rgmf**
**Describe the pull request**
Track deletion is handled by a started/binded service.
The service shows a notification on the status bar with a progress bar so user can use OpenTracks tough tracks are deleting. Nothing is blocked. If deletion is slow then user can see the progress.

Captura de pantalla_2021-05-09_17-20-25

**Link to the the issue**
#735

### CODE REVIEW

File: .idea/misc.xml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <project version="4">
-  <component name="ProjectRootManager" version="2" languageLevel="JDK_1_8" default="true" project-jdk-name="1.8" project-jdk-type="JavaSDK">
+  <component name="ProjectRootManager" version="2" languageLevel="JDK_11" default="true" project-jdk-name="1.8" project-jdk-type="JavaSDK">
```
Comment (dennisguse) [inline .idea/misc.xml:3]: Can you move this to a separate commit? It is related to AndroidStudio 4.2 update.
(incl runConfigurations.xml)

File: src/main/java/de/dennisguse/opentracks/AbstractTrackDeleteActivity.java [hunk 2]
```diff
@@ -38,21 +37,46 @@
...
+    private static final String TAG = AbstractTrackDeleteActivity.class.getSimpleName();
 
     protected static final int GPS_REQUEST_CODE = 6;
-    private static final int DELETE_REQUEST_CODE = 3;
+
+    private TrackDeleteServiceConnection trackDeleteServiceConnection;
+
+    private ArrayList<Track.Id> trackIds;
+
+    private final Runnable bindChangedStatusCallback = new Runnable() {
```
Comment (dennisguse) [inline src/main/java/de/dennisguse/opentracks/AbstractTrackDeleteActivity.java:51]: We should change this interface as the ServiceConnection should only call this when it get's connected.
Also in the other activities this code is not suitable for "service died notifications".
And then we can actually pass the service as a parameter.

File: src/main/java/de/dennisguse/opentracks/AbstractTrackDeleteActivity.java [hunk 3]
```diff
@@ -38,21 +37,46 @@
...
+        @Override
+        public void run() {
+            if (trackDeleteServiceConnection == null) {
```
Comment (dennisguse) [inline src/main/java/de/dennisguse/opentracks/AbstractTrackDeleteActivity.java:54]: should not happen

File: src/main/java/de/dennisguse/opentracks/AbstractTrackDeleteActivity.java [hunk 4]
```diff
@@ -38,21 +37,46 @@
...
     @Override
-    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
-        if (requestCode == DELETE_REQUEST_CODE) {
-            onTrackDeleted();
-        } else {
-            super.onActivityResult(requestCode, resultCode, data);
+    protected void onStop() {
+        super.onStop();
+        if (trackDeleteServiceConnection != null) {
+            trackDeleteServiceConnection.unbind(this);
```
Comment (dennisguse) [inline src/main/java/de/dennisguse/opentracks/AbstractTrackDeleteActivity.java:79]: also `trackDeleteServiceConnection = null`

null check not needed.

File: src/main/java/de/dennisguse/opentracks/AbstractTrackDeleteActivity.java [hunk 5]
```diff
@@ -38,21 +37,46 @@
...
+            }
+
+            service.getDeletingStatusObservable().observe(AbstractTrackDeleteActivity.this, status -> onTrackDeleteStatus(status));
+        }
+    };
+
+    @Override
+    protected void onStart() {
+        super.onStart();
+        if (trackDeleteServiceConnection == null) {
```
Comment (dennisguse) [inline src/main/java/de/dennisguse/opentracks/AbstractTrackDeleteActivity.java:69]: always create a new one

File: src/main/java/de/dennisguse/opentracks/TrackRecordedActivity.java [hunk 6]
```diff
@@ -248,9 +249,16 @@ public boolean onOptionsItemSelected(MenuItem item) {
         return recordingStatus.getTrackId();
     }
 
+    @Override
+    protected void onTrackDeleteStatus(TrackDeleteService.DeleteStatus deleteStatus) {
```
Comment (dennisguse) [inline src/main/java/de/dennisguse/opentracks/TrackRecordedActivity.java:253]: why twice?
onDeleteConfirmed does something very similar, right?

File: src/main/java/de/dennisguse/opentracks/AbstractTrackDeleteActivity.java [hunk 7]
```diff
@@ -69,6 +93,8 @@ protected void deleteTracks(Track.Id... trackIds) {
     public void onConfirmDeleteDone(Track.Id... trackIds) {
```
Comment (dennisguse) [inline src/main/java/de/dennisguse/opentracks/AbstractTrackDeleteActivity.java:93]: Why not pass a list?

File: src/main/java/de/dennisguse/opentracks/services/TrackDeleteService.java [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+        public TrackDeleteService getService() {
+            return TrackDeleteService.this;
+        }
+    }
+
+    public static class DeleteStatus {
+        private int progress;
+        private int max;
+
+        public DeleteStatus(int progress, int max) {
```
Comment (dennisguse) [inline src/main/java/de/dennisguse/opentracks/services/TrackDeleteService.java:142]: private

File: src/main/java/de/dennisguse/opentracks/services/TrackDeleteService.java [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+        }
+
+        public TrackDeleteService getService() {
+            return TrackDeleteService.this;
+        }
+    }
+
+    public static class DeleteStatus {
+        private int progress;
+        private int max;
```
Comment (dennisguse) [inline src/main/java/de/dennisguse/opentracks/services/TrackDeleteService.java:140]: max is count of tracks to be deleted and progress the current one?

File: src/main/java/de/dennisguse/opentracks/services/TrackDeleteServiceConnection.java [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+import androidx.annotation.NonNull;
+
+import java.util.ArrayList;
+
+import de.dennisguse.opentracks.BuildConfig;
+import de.dennisguse.opentracks.content.data.Track;
+
+import static de.dennisguse.opentracks.services.TrackDeleteService.EXTRA_TRACK_IDS;
+
+public class TrackDeleteServiceConnection implements ServiceConnection {
```
Comment (dennisguse) [inline src/main/java/de/dennisguse/opentracks/services/TrackDeleteServiceConnection.java:18]: Same as TrackRecordingServiceConnection, right?
Why not use ServiceConnection<T extends Service> as a base class?

**Review by dennisguse — COMMENTED** [review]
Did not yet test (was too late).

**Comment by dennisguse** [comment]
And on my personal whishlist: an EspressoUI test that checks that deletion works.
Just in case somebody (usually me) tries to "improve" something ;)

File: src/main/java/de/dennisguse/opentracks/AbstractTrackDeleteActivity.java (hunk 2 above)
Comment (rgmf) [reply src/main/java/de/dennisguse/opentracks/AbstractTrackDeleteActivity.java:51]: Okay, I followed the same approach than TrackRecordingService. In fact I loved that approach the first time I saw it :roll_eyes: 
Do you know another approach for this?
I'm going to look into it.

File: src/main/java/de/dennisguse/opentracks/AbstractTrackDeleteActivity.java (hunk 7 above)
Comment (rgmf) [reply src/main/java/de/dennisguse/opentracks/AbstractTrackDeleteActivity.java:93]: I didn't want to touch it.

Comment (rgmf) [reply src/main/java/de/dennisguse/opentracks/AbstractTrackDeleteActivity.java:93]: This should be fixed/changed in other separated issue because it affects a lot of parts: `Track.Id... trackIds` are everywhere.

File: src/main/java/de/dennisguse/opentracks/services/TrackDeleteService.java (hunk 9 above)
Comment (rgmf) [reply src/main/java/de/dennisguse/opentracks/services/TrackDeleteService.java:140]: That's right.
I've used the same names that Android in `BuilderNotification.setProgress(int max, int progress, int indeterminate)`

File: src/main/java/de/dennisguse/opentracks/services/TrackDeleteServiceConnection.java (hunk 10 above)
Comment (rgmf) [reply src/main/java/de/dennisguse/opentracks/services/TrackDeleteServiceConnection.java:18]: > Same as TrackRecordingServiceConnection, right?

Yes, the approach is the same, but the implementation of both connections are different.

**Comment by rgmf** [comment]
> And on my personal whishlist: an EspressoUI test that checks that deletion works.
> Just in case somebody (usually me) tries to "improve" something ;)

Not usually you but everyone :rofl: :rofl: :rofl: 

That makes sense, I should implement some EspressoUI tests... so I'm going to do it.

File: src/main/java/de/dennisguse/opentracks/AbstractTrackDeleteActivity.java (hunk 4 above)
Comment (rgmf) [reply src/main/java/de/dennisguse/opentracks/AbstractTrackDeleteActivity.java:79]: null check if needed because the connection could be unbind before onStop (when deletion service finishes).
Look at line 120.

File: src/main/java/de/dennisguse/opentracks/AbstractTrackDeleteActivity.java (hunk 2 above)
Comment (dennisguse) [reply src/main/java/de/dennisguse/opentracks/AbstractTrackDeleteActivity.java:49]: Creating an interface and implementing on the activity/fragment and pass it to the TrackRecordingService.
(side note: I am working on making TrackController a fragment and might need another callback there aka service disconnected).

**Comment by rgmf** [comment]
@dennisguse Espresso test added to check that a track is delete when user select a item in the list view and click on delete action menu item.

PS/ This afternoon has been funny learning Espresso :smile:  I've never used it an I still have a lot of things to learn. I'll learn more about it because it looks important.

PS1/ I'll try to make an effort to add more tests to OT... It will worth. But it's so boring... :laughing:

**Comment by dennisguse** [comment]
@rgmf Awesome!
i also promise to write some more tests (especially UI)!

**Comment by rgmf** [comment]
It's ready for me but we have a problem with the vacuum when deleting. In a phone I have it takes too long to delete a track because the vacuum.

When that occurs then if you try to do something while deleting an ANR can showed (if that action imply a query) or other not desired behaviour. For example, if you try to show aggregated stats while deleting then they won't show until deleting is ended.

Because of this problem I've added a workaround: avoid all actions until deletion is finished and user sees a screen like this:

imagen

**Comment by rgmf** [comment]
Another possible solution would be create a transparent Dialog and the UI would be like before but the lifecycle would be more complicated :thinking:

**Comment by dennisguse** [comment]
@rgmf Could it be that we are loading in the UI thread and we shouldn't do that?
(I guess, you see the issue mainly in the loaderCallback in the two listActivities)

PS: It reliably gets more complicated than it initially looks like :)

**Comment by rgmf** [comment]
> @rgmf Could it be that we are loading in the UI thread and we shouldn't do that?
> (I guess, you see the issue mainly in the loaderCallback in the two listActivities)
> 
> PS: It reliably gets more complicated than it initially looks like :)

That's right... We are querying database on UI thread. I'm going to re-think it. Because if a query is blocked until vacuum is done then UI shouldn't be freeze.

**Comment by dennisguse** [comment]
@rgmf Do you know at where this is happening?
I just briefly checked the lists and those calls should be in a background thread using LoaderManager.

**Comment by rgmf** [comment]
> @rgmf Do you know at where this is happening?
> I just briefly checked the lists and those calls should be in a background thread using LoaderManager.

It happends when deleting gets stuck in the vavuum and user opens a track. In TrackRecordedActivity: contentProviderUtils.getTrack gets freeze because is in the UI thread and other sql execution is running.

**Comment by dennisguse** [comment]
My proposal: Let's live this (for now) aka let's merge without tackling these issues in this PR.

TL;DR 

I was worried that you would find such things.
I guess, then it even affects more places like MarkerDetailActivity as well as the *EditActivities (while loading and storing).

The question is how lilkely/bad is this?
From my code perspective, it should be okay (for now) as you cleaned up one part and just found some other hidden gems.
And for the user: it shouldn't happen that often, right?
* Deletion is not the daily business
* Depends on actual database size (mainly would affect power users)
* Depends on device performance

**Comment by rgmf** [comment]
We should to do all database operations out of UI thread, right? We could do that with LiveData, what do you think? I can follow up this on this PR for completing it.

I know it's a lot of work but... I can do it... or not... will see :rofl: 

Deleting now is working, it's not hurry for this PR.

**Comment by dennisguse** [comment]
@rgmf That would be awesome!
Do you need help?

PS: I have no trouble merging this PR even it may produce ANRs, as it cleans up other things and makes our live already easier.
